node-wasm/src/ports.rs

```diff
@@ -0,0 +1,250 @@
+use js_sys::{Array, Function, Reflect};
+use serde_wasm_bindgen::{from_value, to_value};
+use tokio::select;
+use tokio::sync::{mpsc, Mutex};
+use tracing::{error, info, trace};
+use wasm_bindgen::closure::Closure;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::{JsCast, JsValue};
+use wasm_bindgen_futures::spawn_local;
+use web_sys::{MessageEvent, MessagePort};
+
+use crate::commands::{NodeCommand, WorkerResponse};
+use crate::error::{Context, Error, Result};
+
+const REQUEST_MULTIPLEXER_COMMAND_CHANNEL_SIZE: usize = 64;
+
+// Instead of just supporting communicaton with just `MessagePort`, allow using any object which
+// provides compatible interface
+#[wasm_bindgen]
+extern "C" {
+    pub type MessagePortLike;
+
+    #[wasm_bindgen (catch , method , structural , js_name = postMessage)]
+    pub fn post_message(this: &MessagePortLike, message: &JsValue) -> Result<(), JsValue>;
+}
+
+impl From<MessagePort> for MessagePortLike {
+    fn from(port: MessagePort) -> Self {
+        JsValue::from(port).into()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientId(usize);
+
+struct ClientConnection {
+    port: MessagePortLike,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl ClientConnection {
+    fn new(
+        id: ClientId,
+        object: JsValue,
+        forward_to: mpsc::Sender<(ClientId, Result<NodeCommand, TypedMessagePortError>)>,
+    ) -> Result<Self> {
+        let onmessage = Closure::new(move |ev: MessageEvent| {
+            let response_tx = forward_to.clone();
+            spawn_local(async move {
+                let message: Result<NodeCommand, _> =
+                    from_value(ev.data()).map_err(TypedMessagePortError::FailedToConvertValue);
+
+                if let Err(e) = response_tx.send((id, message)).await {
+                    error!("message forwarding channel closed, shouldn't happen: {e}");
+                }
+            })
+        });
+
+        let port = prepare_message_port(object, &onmessage)
+            .context("failed to setup port for ClientConnection")?;
+
+        Ok(ClientConnection {
+            port,
+            _onmessage: onmessage,
+        })
+    }
+
+    fn send(&self, message: &WorkerResponse) -> Result<()> {
+        let message_value = to_value(message).context("could not serialise message")?;
```

Comment:
So in extension template we currently need to sanitize messages.
```
// https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Chrome_incompatibilities#data_cloning_algorithm
// Chrom* uses JSON serialisation algoritm for messges sent over runtime.Port, which strips keys with a value of `undefined`.
// This mangles some of lumina responses, e.g. successful start node response is `{ NodeStarted: { Ok : undefined } }`, which
// then becomes `{ NodeStarted: {} }`.
// Thankfully, replacing `undefined` with `null` works as a workaround, since we don't care about key value anyway
```

We could migrate this comment and solve it here by using
```rs
use serde_wasm_bindgen::Serializer;
use serde::Serialize;
 ...
let serializer = Serializer::default().serialize_missing_as_null(true);
let message_value = message.serialize(&serializer).context("could not serialise message")?;
```

Reply:
Nice, didn't know about it! Switched to using `Serializer::json_compatible`